### ISSUE TITLE
sbt blocks several ScalaTest features by withholding event reports

### ISSUE THREAD

**Issue by bvenners** [sbt/sbt#2401]
Hi,

At some point sbt started holding onto events until a test class completes, then emitting them all at once. This may have made the output of other test frameworks better, but it made ScalaTest output worse. ScalaTest already outputs its events in such a way that it balances the user seeing the output for each test class together with the desire of the user to see that progress is happening. It will show the output of one test class, then the next, then the next, like sbt does, but unlike sbt as far as I know, ScalaTest will eventually time out and switch to the next test class. That way if one test takes a long time, or hangs forever, eventually ScalaTest will move on and start outputting events from another test class. As I mentioned, this balances the user desire to see each test class reported together with the desire to see progress when one test is slow or hanging.

Whats' worse is that some of ScalaTest's events are designed to be emitted immediately. These are called "notes" and "alerts". These allow users to get status updates, or warnings of problems, as a long-running test proceeds. Some users have tests that run for three minutes, so these updates are desirable to them to make sure tests are proceeding.

Lastly, occasionally tests hang forever for users because of deadlocks or other issues. ScalaTest has a nice feature, called "slowpoke notifications," to help users figure out what test is hanging. With this enabled, an alert is fired if a test has been running more than a certain amount of time. The alert indicates the suite and test name, so you can find out which test or tests are hanging. It is a great feature, but doesn't work on sbt anymore, because sbt refuses to show the event until the suite has finished. If the suite is hanging, it won't finish, so you can't use it to find out what's hanging. On the JVM you can often do a thread dump, because ScalaTest puts the suite and test name in the thread name, but that doesn't work on Scala.js.

I know other test frameworks don't have these features, but it would be good to either have a way to shut it off, or to automatically shut it off for ScalaTest.

Thanks.

Bill

**Comment by eed3si9n** [comment]
See http://www.scala-sbt.org/0.13/docs/Testing.html#Output

Could you try if `logBuffered in Test := false` works for you?

**Comment by bvenners** [comment]
I will try that right now...

**Comment by bvenners** [comment]
That works! I'll add it to our documentation asap, as I know users who know sbt very well who have been complaining about this pain point.

But I wonder, do you think there's someway to make that the default at least for ScalaTest? Given I didn't find this in the sbt documentation for maybe a year, it is possible users won't find it in ScalaTest documentation. Hmm. That might complicate sbt to have different defaults for different test frameworks, but as far as I know for the other test frameworks the existing default is probably appropriate.

**Comment by eed3si9n** [comment]
> That might complicate sbt to have different defaults for different test frameworks, but as far as I know for the other test frameworks the existing default is probably appropriate.

I wasn't there when that decision was made (c8fe1a3c1d32d98a3ecfd84547fffe8db9b83c25 ?), and I have no idea if this behavior benefits any testing frameworks or tasks. Potential reasons:
- performance tradeoff of file I/O?
- prevention of interleaving outputs from multiple tests running in parallel?

**Comment by bvenners** [comment]
2011? I wonder in what version of sbt that came out. 

ScalaTest already takes care of the interleaving of multiple tests running in parallel, as I mentioned above. But in a slightly better way I think, because it can eventually time out and moves on. I am not aware that any other test framework does that, so for the others, this probably makes sense as the default in sbt. But I don't know from direct experience.  Be interesting to see what happens in a specs2 or JUnit project when this is disabled and parallel execution is performed.

**Comment by eed3si9n** [comment]
> 2011? I wonder in what version of sbt that came out.

The commit shows all the tags. The earliest is v0.9.5, but I think sbt sorted of jumped from 0.7 to 0.10 around the time if I remember correctly.

Like other testing related things in sbt, it innovated ahead of the time, and filled in the gap. Now that testing libraries are managing parallelism or logging or whatever, I'd be happy to change things around if you, @etorreborre, and @benmccann think it makes sense to do so.

**Comment by bvenners** [comment]
I'm not sure it matters. I can just document to set logBuffered to false when using ScalaTest and see how that goes. The only thing that could potentially be done, perhaps, is to change sbt's default to the opposite. That seems like a fairly drastic break to existing builds. Nevertheless we'll investigate to see what it does to JUnit, specs2, and ScalaCheck tests and I'll post here.

**Comment by etorreborre** [comment]
I remember being told by Mark Harrah to set `logBuffer` to `false` a long time ago, as I started running things concurrently. This is in [specs2 build](https://github.com/etorreborre/specs2/blob/master/project/build.scala#L205) and I mention it in the [docs](https://etorreborre.github.io/specs2/guide/SPECS2-3.7/org.specs2.guide.Runners.html).

As far as I know the main downside is the interleaving of traces when running several classes in parallel. In my experience this is not really a problem because I usually want to see the full output when I'm running a single class and just failures/errors if I'm running a suite.

**Comment by bvenners** [comment]
Hi Eric,

That's interesting. Sounds like maybe false would have been a better default for logBuffered. We have yet to try false with other test frameworks. We'll do that this week. 

Bill

**Comment by benmccann** [comment]
Yeah, I wouldn't know off the top of my head how it would impact the junit output and we'd have to test it out to see. Btw, if we're already going to be testing the junit output then it might be a good time to finally resolve https://github.com/sbt/junit-interface/issues/62, which I've been consistently too lazy to do

**Comment by bvenners** [comment]
Hi Ben,

I think making -a the default is something that would be done in sbt/junit-interface, not sbt itself. I agree it makes sense as a better default.

Bill
